# NanoHunter

## <a name="nh"></a>What is NanoHunter ?
NanoHunter is an analysis pipeline designed for long-read single-cell RNA-seq data.
It mainly consists of two part: 
1. Barcode & UMI calling
2. Downstream analysis: cell type-specific splicing and allele-specific splicing

<!-- ![workflow](https://github.com/Xinglab/NanoHunter/blob/master/NanoHunter_workflow.png) -->


## <a name="install"></a>Installation
### <a name="os"></a>Operating system and python version
NanoHunter was written in python3 and tested on Linux/Unix systems, so it may not work well with python2 and/or other systems(Windows/Mac).

<!-- ### <a name="dep"></a>~~Dependencies~~ installed via conda
* [python3](https://www.python.org/) >= 3.8
* [pysam](https://pysam.readthedocs.io/en/latest/) >= 0.21.0
* [edlib](https://pypi.org/project/edlib/) >= 1.3.9
* [kneed](https://pypi.org/project/kneed/) >= 0.8.3
* [scipy](https://scipy.org/) >= 1.10.1
* [rpy2](https://github.com/rpy2/rpy2) >= 3.5.11
* [PyVCF3](https://github.com/dridk/PyVCF3) >= 1.0.3
* [pyinterval](https://pyinterval.readthedocs.io/en/latest/) >= 1.2.0
* [duckdb](https://duckdb.org/docs/api/python/overview.html) >= 0.7.1
* [tabix](https://anaconda.org/bioconda/tabix)
* [whatshap](https://whatshap.readthedocs.io/en/latest/index.html) >= 1.7

(Optional) For rolling circle amplification (RCA) single-cell long-read data:
* [TideHunter](https://github.com/yangao07/TideHunter) >= 1.5.4 -->

### <a name="conda"></a>Via conda
```
conda create -n nanohunter python=3.8 nanohunter
conda activate nanohunter
```

### <a name="conda"></a>Via pip
```
pip install nanohunter
```

### <a name="source"></a>From source files
```
git clone 
cd 
pip install .
```

##  <a name="consensus"></a>(Optional) Consensus calling for RCA single-cell long-read data with [TideHunter(≥1.5.4)](https://github.com/yangao07/TideHunter)
### Input
* `rca_long_read.fq`: RCA long-read fasta/fastq file
* `five_prime.fa`, `three_prime.fa`: 5' and 3' sequence of RCA library. If splint sequence was used, please split splint sequence into two halves and use them as 5' and 3' sequence. Note that both 5' and 3' sequence need to have orientation of 5'->3', i.e., 3' sequence need to be reverse-complement

### Command
```
TideHunter rca_long_read.fq \
           -5 five_prime.fa \
           -3 three_prime.fa \
           -lF -t 16 \
           -o consensus.fa
```

## <a name="bu_calling"></a> 1. Barcode & UMI calling
NanoHunter identifies barcode and UMI from sorted alignment BAM file of single-cell long reads ***without*** using reference barcode from short-read data. 

For RCA long reads, before barcode/UMI calling, consensus sequences need to be generated (see [above](#consensus)) and then mapped to reference genome.
### 1.1 Input 
* Required:
  * `long_read.sorted.bam`: sorted long-read BAM (recommend using [minimap2](https://github.com/lh3/minimap2))
  * `read_isoform_compatible.tsv`: tabular file of compatible isoforms of each read, generated by [ESPRESSO(≥1.3.1)](https://github.com/Xinglab/espresso) 
  * `updated.gtf`: updated GTF annotation, generated by [ESPRESSO(≥1.3.1)](https://github.com/Xinglab/espresso)
  * output folder
* Optional:
  * `cell_barcode.tsv`: cell barcode list, nanohunter will directly use it to guide the barcode calling, only long reads with cell barcode in the list will be kept
  * `annotation.gtf`: annotation GTF file, to retrieve gene names. If not provided, NanoHunter will use gene ids as gene names
  * barcode sequence length (default: 16)
  * max. allowed edit distance between barcode and reference barcode (default: 2)
  * UMI sequence length (default: 12)
  * max. allowed edit distance between PCR-duplicated UMIs (default: 1)

### 1.2 Command
```
nanohunter long_read.sorted.bam \
           updated.gtf          \
           read_isoform_compatible.tsv \
           -g annotation.gtf  \
           output foler
```
### 1.3 Output
* `bc_umi.bam`: BAM file with barcode/UMI information for each long read, BAM tags: `CB` and `UB`, for barcode and UMI.  Note that only long read with barcode/UMI called are kept
* `bc_umi.tsv`: tabular file with barcode/UMI/gene/transcript information for all barcode-called long reads
* `expression_matrix/`: folder containing 10X format sparse expression matrix at both gene and transcrpt level, can be directly parsed using standard single-cell analysis tools, like [Seurat](https://satijalab.org/seurat/)/[Azimuth](https://satijalab.github.io/azimuth/articles/run_azimuth_tutorial.html)
* `ref_bc.tsv`: reference cell barcode list identified from long reads. Note that if cell barcode list was provided to NanoHunter, `ref_bc.tsv` file will not be generated
* `high_qual_bc_umi.rank.tsv`: cell barcode list ranked by unique UMI count based on all high-quality long reads. Note that if cell barcode list was provided to NanoHunter, `high_qual_bc_umi.rank.tsv` file will not be generated
  
Format of `bc_umi.tsv`:

| cell barcode | UMI | # reads | compatible transcript id | gene id | gene name | read names (separeted by \`,\`)|
|-|-|-|-|-|-|-|
| ATCACGACACTTTAGG | ATCACATCCATG | 3 | ENST00000407249, ENST00000341832 | ENSG00000248333 | CDK11B | 741aa2c2-5840-4a29-bd90-3bdcb71604ba, 05f8432a-7f7e-446d-a6d5-8ab9e4eb5102, 6bb13ee1-7397-435c-8840-aeb2a90cf4ab |


## <a name="cell_clutering"></a>2. Cell clustering and annotating
All the downstream single-cell long-read analysis rely on the cell type clustering result, which could be acomplished by running [Seurat](https://satijalab.org/seurat/) on the the gene expression matrix and annotating the cell clusters manually or based on known reference annotation like [Azimuth](https://satijalab.github.io/azimuth/articles/run_azimuth_tutorial.html).

For example, for human peripheral blood mononuclear cells (PBMC), the clustering and annotating result can be obtained by mapping the Azimuth human PBMC reference dataset. 

We provide the [R script](blob/main/scripts_for_paper/README.md#run_azimuth) for human PBMC data. For data from other species/tissues, this needs to be done manually by users.

## <a name="cell_splice"></a>3. Cell type-specific splicing analysis
NanoHunter performs ***pairwise comparison*** to identify differntially spliced genes/transcripts between each two cell types/clusters.
### 3.1 Input
* Required
  * `expression_matrix/transcript`: transcript count matrix directory, generated by NanoHunter
  * `bc_to_cell_type.tsv`: list of barcode and corresponding cell type/cluster, generated by [Seurat](https://satijalab.org/seurat/)/[Azimuth](https://satijalab.github.io/azimuth/articles/run_azimuth_tutorial.html)
  * `out_prefix`: prefix of output files
* Optional
  * min. number of reads to keep a transcript for each cell type (default: 10)
  * min. number of transcripts to keep a gene for each cell type (default: 2)
  * list of genes of interest. If provided, all genes will still be processed, but only differentially spliced genes that are in this list will be output
### 3.2 Command
```
nh_cell_type_specific_splicing expression_matrix/transcript \
                               bc_to_cell_type.tsv            \
                               out_prefix
```

### 3.3 Output
* `*_diff_splice_genes.tsv`: differentially spliced genes, FDR ≤0.05, max. delta ratio ≥0.05
* `*_diff_splice_transcripts.tsv`: differentially spliced transcripts, gene FDR ≤0.05, transcript p value ≤0.05, detal ratio ≥0.05

Format of `*_diff_splice_genes.tsv`:

| cell_type1| cell_type2|gene_id| gene_name| gene_fdr| max_delta_ratio| transcript_ids| cell1_counts| cell2_counts| cell1_ratios| cell2_ratios|
|-|-|-|-|-|-|-|-|-|-|-|
|Epithelial| Mesenchymal|ENSG00000026508| CD44| 7.96e-65| 0.68| ENST00000263398, ENST00000433892, ENST00000527326, ESPRESSO:chr11:443:2, ESPRESSO:chr11:443:3| 61.43,122.97, 20.0,23.41,13.63| 322.74,4.24,15.0,1.33-05,1.33e-05| 0.25,0.50,0.08,0.09,0.05| 0.94,0.01,0.04,3.89e-08,3.89e-08 |

Format of `*_diff_splice_transcripts.tsv`:

| cell_type1| cell_type2 | gene_id| gene_name| transcript_id| gene_fdr| transcript_p| delta_ratio| count1| count2| ratio1| ratio2|
|-|-|-|-|-|-|-|-|-|-|-|-|
| Epithelial| Mesenchymal| ENSG00000026508| CD44| ENST00000263398| 7.96e-65| 8.25e-73| 0.68| 61.43| 322.74| 0.25| 0.94|



## <a name="allele_splice"></a>4. Allele-specific splicing analysis
With additional whole-genome sequencing data, NanoHunter can identify allele-specific splicing within each cell type.
The allele-specific splicing analysis mainly consists of three steps:
1. Phasing long reads with whatshap
2. Identify allele-specific spliced genes/transcripts within each cell type
3. Search for disease-associated GWAS SNPs from identified allele-specific spliced genes

### <a name="whatshap"></a>4.1 Phasing long reads with `whatshap`
* Input
  * `wgs_phased.vcf`: WGS phased VCF file, generated from WGS short-read data using [GATK](https://gatk.broadinstitute.org/hc/en-us) and [SHAPEIT](https://odelaneau.github.io/shapeit5/)
  * `ref.fa`: reference genome fasta file
  * `bc_umi.bam`, BAM file with barcode/UMI information, generated by NanoHunter
* Command
```
whatshap haplotag wgs_phased.vcf bc_umi.bam \
                  --reference ref.fa        \
                  --ignore-read-groups      \
                  --skip-missing-contigs    \
                  -o bc_umi_hap.bam         \
                  --output-haplotag-list hap_list.tsv
```
* Output
  * `bc_umi_hap.bam`: BAM file with barcode/UMI and additional haplotype information
  * `hap_list.tsv`: tabular file containing haplotype information for barcode-called long reads
  
### <a name="allele_splice_32"></a>4.2 Identifiy allele-specific splicing
* Input
  * `bc_umi.tsv`: tabular file with barcode/UMI/gene/transcript information, generated by NanoHunter
  * `hap_list.tsv`: tabular file containing haplotype information, generated by [whatshap](https://whatshap.readthedocs.io/en/latest/index.html), see [above](#whatshap)
  * `bc_to_cell_type.tsv`: list of barcode and corresponding cell type/cluster, generated by [Seurat](https://satijalab.org/seurat/)/[Azimuth](https://satijalab.github.io/azimuth/articles/run_azimuth_tutorial.html) 

* Command
```
nh_allele_specific_splicing bc_umi.tsv \
                            bc_to_cell_type.tsv \
                            hap_list.tsv \
                            output_prefix

```
* Output
  * `*_allele_splice_gene_detailed.tsv`: allele-specific spliced genes, FDR ≤0.05, max. delta ratio ≥0.05
  * `*_allele_splice_transcript_detailed.tsv`: allele-specific spliced transcripts, gene FDR ≤0.05, transcript p value ≤0.05, delta ratio ≥0.05
  
Format of `*_allele_splice_gene_detailed.tsv`:
|cell_type| gene_id| gene_name| gene_fdr| transcripts| hap1_counts| hap2_counts| hap1_ratios| hap2_ratios|
|-|-|-|-|-|-|-|-|-|
|Mono| ENSG00000105383| CD33| 4.42e-07| ENST00000262262, ENST00000421133| 58,9| 24,41| 0.86,0.13| 0.36,0.63|

Format of `*_allele_splice_transcript_detailed.tsv`:
|cell_type| transcript_id| transcript_p| gene_id| gene_name| gene_fdr| hap1_count| hap2_count| hap1_ratio| hap2_ratio|
|-|-|-|-|-|-|-|-|-|-|
|Mono| ENST00000262262| 3.58e-09| ENSG00000105383| CD33| 4.42e-07| 58| 24| 0.86| 0.36|

### 4.3 Identify disease-associated GWAS SNPs from allele-specific spliced genes
* Input
  * `*_allele_splice_gene_detailed.tsv`: allele-specific spliced genes, generated by [nh_allele_specific_splicing](#allele_splice_32)
  * `wgs_phased.vcf`: WGS phased VCF file, generated from WGS short-read data using [GATK](https://gatk.broadinstitute.org/hc/en-us) and [SHAPEIT](https://odelaneau.github.io/shapeit5/)
  * `annotation.gtf`: annotation GTF file
  * `gwas_catalog_efo.tsv`: GWAS catalog database with EFO term
  * `ld.duckdb`: linkage disequilibrium (LD) score database in duckdb format

* Command
```
nh_gene_with_gwas_disease_snp -g allele_splice_gene_detailed.tsv \
                              annotation.gtf       \
                              wgs_phased.vcf       \
                              gwas_catalog_efo.tsv \
                              ld.duckdb            \
                              output_prefix
```

* Output
  * `*_allele_splice_gwas_detailed.tsv`: GWAS trait and LD information for each SNP of allele-specific spliced gene
  * `*_allele_splice_gwas_efo_term.tsv`: GWAS disease-associated EFO term of allele-specific spliced gene 
  
Format of `*_allele_splice_gwas_detailed.tsv`:
|snp_id| gene_name| gene_id| haplotype| gwas_trait| gwas_parent_term| ld_trait| ld_parent_term| ld_snps| ld_scores|
|-|-|-|-|-|-|-|-|-|-|
|rs3865444| CD33| ENSG00000105383| H2| Alzheimer disease| Neurological disorder| NA, Alzheimer disease, late-onset Alzheimers disease, NA, NA| NA, Neurological disorder,Neurological disorder, NA, NA| rs12459419, rs7245846, rs1354106, rs34813869, rs33978622| 1.0,0.89,0.88,0.88,0.96|

Format of `*_allele_splice_gwas_efo_term.tsv`:
|gene_name| gene_id| Cancer| Immune system disorder| Neurological disorder| Cardiovascular disease| Digestive system disorder| Metabolic disorder| Response to drug| Other disease|
|-|-|-|-|-|-|-|-|-|-|
|CD33| ENSG00000105383| False| False | True | False| False| False| False| False|